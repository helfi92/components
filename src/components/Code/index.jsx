import React, { Component } from 'react';
import { string } from 'prop-types';
import { getLanguage, highlight } from 'highlight.js';
import classNames from 'classnames';
import 'highlight.js/styles/atom-one-dark.css';

const validLanguage = (props, propName) => {
  const language = props[propName];

  if (!getLanguage(language)) {
    return new Error(`Language '${language}' not supported by highlight.js`);
  }
};

/**
 * Render children as syntax-highlighted monospace code.
 */
export default class Code extends Component {
  static propTypes = {
    /**
     * The content to render as highlighted syntax.
     */
    children: string.isRequired,
    /**
     * A highlight.js language identifier.
     */
    language: validLanguage.isRequired,
    /** The CSS class name of the wrapper element */
    className: string,
  };

  static defaultProps = {
    className: null,
  };

  state = {
    code: null,
  };

  componentDidMount() {
    const { children, language } = this.props;

    this.setState({
      code: highlight(language, children, true).value,
    });
  }

  render() {
    const { language, className, ...props } = this.props;
    const { code } = this.state;

    /* eslint-disable react/no-danger */
    return (
      <pre className={classNames(`language-${language}`, className)} {...props}>
        {code && <code dangerouslySetInnerHTML={{ __html: code }} />}
      </pre>
    );
    /* eslint-enable react/no-danger */
  }
}
