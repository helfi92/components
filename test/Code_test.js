import React from 'react';
import Code from '../src/components/Code';

describe('<Code />', () => {
  it('should render correctly', () => {
    const text = JSON.stringify({ foo: 'bar' });
    const code = mount(
      <Code language="json">{text}</Code>
    );

    expect(code).toMatchSnapshot();
  });
});
