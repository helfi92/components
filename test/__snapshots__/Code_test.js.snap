// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Code /> should render correctly 1`] = `
<Code
  className={null}
  language="json"
>
  <pre
    className="language-json"
  >
    <code
      dangerouslySetInnerHTML={
        Object {
          "__html": "{<span class=\\"hljs-attr\\">\\"foo\\"</span>:<span class=\\"hljs-string\\">\\"bar\\"</span>}",
        }
      }
    />
  </pre>
</Code>
`;
